File: Gryllus/scripts/load_data.R
library(Matrix)
library(Seurat)
library(dplyr)
library(readr)
library(stringr)

# -------------------------------
# Step 1: Load raw Visium data
# -------------------------------

# Define paths
base_path <- "../../datashare/grillus"
matrix_path <- file.path(base_path, "outs", "filtered_feature_bc_matrix")
spatial_path <- file.path(base_path, "outs", "spatial")
data_path <- file.path(base_path, "files")

# Load matrix and metadata
expr_matrix <- readMM(file.path(matrix_path, "matrix.mtx.gz"))
barcodes <- read_tsv(file.path(matrix_path, "barcodes.tsv.gz"), col_names = "barcode")
features <- read_tsv(file.path(matrix_path, "features.tsv.gz"), 
                     col_names = c("gene_id", "gene_name", "feature_type"))

# Clean gene IDs
features <- as.data.frame(features)
features <- features %>%
  mutate(
    gene_id = str_replace_all(gene_id, "\t|Gene Expression", "") %>%
      str_trim() %>%
      str_to_lower() %>%
      str_sub(1, 9)
  )

# Apply consistent gene IDs to matrix
expr_matrix <- as(expr_matrix, "dgCMatrix")
rownames(expr_matrix) <- make.unique(features$gene_id)
colnames(expr_matrix) <- barcodes$barcode

cat("Matrix loaded with dimensions:", dim(expr_matrix)[1], "genes x", dim(expr_matrix)[2], "barcodes\n")

# -------------------------------
# Step 2: Create Seurat object
# -------------------------------
seurat_obj <- CreateSeuratObject(counts = expr_matrix, assay = "Spatial")

# -------------------------------
# Step 3: Load gene signature files
# -------------------------------
gene_list <- read_csv(file.path(data_path, "liste_gene_and_code.csv"))
genes_growth <- read_csv(file.path(data_path, "de_gene_names_growth.csv"))
genes_regen_up <- read_csv(file.path(data_path, "ur_regeneration.csv"))
genes_regen_down <- read_csv(file.path(data_path, "dr_regeneration.csv"))
genes_scarring <- read_csv(file.path(data_path, "gene_name_scaring.csv"))

# Clean gene names
clean_genes <- function(df) {
  df[[1]] <- df[[1]] %>% 
    str_replace_all("\t", "") %>% 
    str_trim() %>% 
    str_to_lower()
  df
}

genes_growth <- clean_genes(genes_growth)
genes_regen_up <- clean_genes(genes_regen_up)
genes_regen_down <- clean_genes(genes_regen_down)
genes_scarring <- clean_genes(genes_scarring)

# -------------------------------
# Step 4: Add metadata to features
# -------------------------------
features_fixed <- features
rownames(features_fixed) <- rownames(seurat_obj)  # Ensure match

features_fixed$growth_gene <- rownames(seurat_obj) %in% genes_growth[[1]]
features_fixed$regen_up_gene <- rownames(seurat_obj) %in% genes_regen_up[[1]]
features_fixed$regen_down_gene <- rownames(seurat_obj) %in% genes_regen_down[[1]]
features_fixed$scarring_gene <- rownames(seurat_obj) %in% genes_scarring[[1]]

seurat_obj@misc$gene_annotations <- features_fixed
gene_annot <- seurat_obj@misc$gene_annotations

# -------------------------------
# Step 5: Load spatial metadata
# -------------------------------
positions <- read_csv(file.path(spatial_path, "tissue_positions.csv"))
colnames(positions)[1] <- "barcode"

# Merge with Seurat object metadata
seurat_obj@meta.data$barcode <- rownames(seurat_obj@meta.data)
seurat_obj@meta.data <- left_join(seurat_obj@meta.data, positions, by = "barcode")


# -------------------------------
# Step 6: Save object
# -------------------------------
saveRDS(seurat_obj, file = file.path(base_path, "grillus_seurat.rds"))

File: Gryllus/scripts/load_data.py
import argparse
import anndata
import pandas as pd
import scipy.io
from scipy.io import mmread

def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--outs", required=True, help="Space Ranger outs directory (contains filtered_feature_bc_matrix and spatial)")
    ap.add_argument("--files", required=True, help="Directory with signature CSVs (liste_gene_and_code.csv etc.)")
    ap.add_argument("--out_h5ad", required=True, help="Output .h5ad path")
    args = ap.parse_args()

    matrix_path  = f"{args.outs}/filtered_feature_bc_matrix"
    spatial_path = f"{args.outs}/spatial"
    data_path    = args.files

    matrix = scipy.io.mmread(f"{matrix_path}/matrix.mtx.gz").tocsc()
    barcodes = pd.read_csv(f"{matrix_path}/barcodes.tsv.gz", header=None, names=["barcode"])
    features = pd.read_csv(f"{matrix_path}/features.tsv.gz", header=None, names=["gene_id", "gene_name", "feature_type"])

    genes_growth    = pd.read_csv(f"{data_path}/de_gene_names_growth.csv")
    genes_regen_up  = pd.read_csv(f"{data_path}/ur_regeneration.csv")
    genes_regen_down= pd.read_csv(f"{data_path}/dr_regeneration.csv")
    genes_scarring  = pd.read_csv(f"{data_path}/gene_name_scaring.csv")

    features['gene_id'] = (features['gene_id']
        .astype(str)
        .str.replace(r'\t', '', regex=True)
        .str.replace('Gene Expression', '', regex=True)
        .str.strip().str.lower()
        .str.slice(0, 9)
    )

    for df in [genes_growth, genes_regen_up, genes_regen_down, genes_scarring]:
        df[df.columns[0]] = (df[df.columns[0]].astype(str)
            .str.replace(r'\t', '', regex=True)
            .str.strip().str.lower()
        )

    adata = anndata.AnnData(X=matrix.T, obs=barcodes, var=features.set_index('gene_id'))
    adata.var.index = adata.var.index.str.slice(0, 9)

    # Nota: tu código tenía un typo: growth_gene usaba genes_scarring
    adata.var['growth_gene']     = adata.var.index.isin(genes_growth[genes_growth.columns[0]].tolist())
    adata.var['regen_up_gene']   = adata.var.index.isin(genes_regen_up[genes_regen_up.columns[0]].tolist())
    adata.var['regen_down_gene'] = adata.var.index.isin(genes_regen_down[genes_regen_down.columns[0]].tolist())
    adata.var['scarring_gene']   = adata.var.index.isin(genes_scarring[genes_scarring.columns[0]].tolist())

    tissue_positions = pd.read_csv(f"{spatial_path}/tissue_positions.csv")
    adata.obs = adata.obs.merge(tissue_positions, on="barcode", how="left")

    adata.write_h5ad(args.out_h5ad)

if __name__ == "__main__":
    main()